### ISSUE TITLE
[FSDP] Sharding params within a node, and gradient/optimizer globally

### ISSUE THREAD

**Issue by ChrisLiu6** [pytorch/pytorch#117968]
### 🚀 The feature, motivation and pitch

Parameter all-gathering is a very frequently called operation with zero3, but gathering across nodes can be slow, which makes zero3 much slower than zero2. When training models like llama 70B, the current HSDP implementation, which shards are things only within nodes, will cause OOM; on the other hand, the FULL_SHARD strategy causes very low throughput. Currently, we have to involve intra-node 8 way tensor parallel + inter-node zero2 to deal with such situation, but this adds the complexity and still leaves the redundancy (e.g. the duplicate activations among tensor parallel ranks if sequence parallel is not further applied).

I guess a different hybrid sharding strategy, namely to shard params within a node, and gradient/optimizer globally, may well solve such problems. For example, given 16 gpus on 2 machines, after all-gathering params globally, rank0 may retain the parameter shard for rank0 and rank8 after resharding.

I am not sure if I've missed anything that makse the solution not as good as I think. Thx

cc @zhaojuanmao @mrshenli @rohan-varma @awgu @fegin @penguinwu @kwen2501

**Comment by ChrisLiu6** [comment]
Ah I found it in zero++, called Hierarchical Partitioning for ZeRO (https://arxiv.org/pdf/2306.10209.pdf) Is there any plan for FSDP to support this?

**Comment by awgu** [comment]
Thanks for filing this issue!

We are working on a new FSDP implementation (https://github.com/pytorch/pytorch/issues/114299). (I am opening PRs these days actually 😅 .) When that is ready for broader adoption (targeting mid-year 2024), it should have this feature.
